File: 3291-find-if-array-can-be-sorted/3291-find-if-array-can-be-sorted.cpp
class Solution {
public:

    bool haveSameSetBits(int a, int b) {
        return __builtin_popcount(a) == __builtin_popcount(b);
    }

    bool canSortArray(vector<int>& nums) {
        
        int N = nums.size(), times = nums.size();
        while (times--) {
            for (int i = 0; i < N - 1; i++) {
                if (haveSameSetBits(nums[i], nums[i + 1]) && nums[i + 1] < nums[i]) {
                    swap(nums[i], nums[i + 1]);
                }
            } 
        }

        return is_sorted(nums.begin(), nums.end());
    }
};

File: 3291-find-if-array-can-be-sorted/3291-find-if-array-can-be-sorted.java
class Solution {
    public boolean canSortArray(int[] nums) {
        int[]ans=new int[nums.length];
        for(int i=0;i<nums.length;i++){
            ans[i]=nums[i];
        }
        Arrays.sort(ans);
        if(Arrays.equals(ans,nums))return true;
        for(int i=0;i<nums.length-1;i++){
            for(int j=0;j<nums.length-1-i;j++){
                if(Integer.bitCount(nums[j])==Integer.bitCount(nums[j+1])&&nums[j]>nums[j+1]){
                    int t=nums[j];
                    nums[j]=nums[j+1];
                    nums[j+1]=t;
                } 
                  
            }
        }
        return Arrays.equals(nums,ans);
    
    }
}